Tests are broken on master for `pypy` / `pypy3`

seems we don't test them in CI any more? perhaps an oversight?

appears to have regressed in #1310

The core of the problem is that `tox -e py27` might match a `python2.7` that is `pypy2` which fails the interpreter matching code (`python` != `pypy`)

These are the failing tests:

```
FAILED tests/unit/interpreters/test_interpreters.py::test_tox_get_python_executable - assert None
FAILED tests/unit/config/test_config.py::test_config_current_py - AssertionError: 1
FAILED tests/integration/test_package_int.py::test_package_setuptools - AssertionError: 1
FAILED tests/unit/test_z_cmdline.py::test_notest - AssertionError: 1
```

They each fail similar to:

```
E       AssertionError: 1
E       Traceback (most recent call last):
E         File "/tmp/tox/.tox/pypy/site-packages/tox/_pytestplugin.py", line 126, in run
E           tox.session.main([str(x) for x in argv])
E         File "/tmp/tox/.tox/pypy/site-packages/tox/session/__init__.py", line 68, in main
E           exit_code = session.runcommand()
E         File "/tmp/tox/.tox/pypy/site-packages/tox/session/__init__.py", line 192, in runcommand
E           return self.subcommand_test()
E         File "/tmp/tox/.tox/pypy/site-packages/tox/session/__init__.py", line 208, in subcommand_test
E           venv.package = self.hook.tox_package(session=self, venv=venv)
E         File "/tmp/tox/.tox/pypy/site-packages/pluggy/hooks.py", line 289, in __call__
E           return self._hookexec(self, self.get_hookimpls(), kwargs)
E         File "/tmp/tox/.tox/pypy/site-packages/pluggy/manager.py", line 87, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/tmp/tox/.tox/pypy/site-packages/pluggy/manager.py", line 81, in <lambda>
E           firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
E         File "/tmp/tox/.tox/pypy/site-packages/pluggy/callers.py", line 208, in _multicall
E           return outcome.get_result()
E         File "/tmp/tox/.tox/pypy/site-packages/pluggy/callers.py", line 81, in get_result
E           _reraise(*ex)  # noqa
E         File "/tmp/tox/.tox/pypy/site-packages/pluggy/callers.py", line 187, in _multicall
E           res = hook_impl.function(*args)
E         File "/tmp/tox/.tox/pypy/site-packages/tox/package/__init__.py", line 16, in tox_package
E           session.package, session.dist = get_package(session)
E         File "/tmp/tox/.tox/pypy/site-packages/tox/package/__init__.py", line 29, in get_package
E           package = acquire_package(config, session)
E         File "/tmp/tox/.tox/pypy/site-packages/tox/package/__init__.py", line 40, in acquire_package
E           path = build_package(config, session)
E         File "/tmp/tox/.tox/pypy/site-packages/tox/package/builder/__init__.py", line 9, in build_package
E           return build(config, session)
E         File "/tmp/tox/.tox/pypy/site-packages/tox/package/builder/isolated.py", line 31, in build
E           raise package_venv.status
E       InterpreterNotFound: InterpreterNotFound: python
E       
E       using tox.ini: /tmp/pytest-of-asottile/pytest-56/popen-gw6/test_package_setuptools0/magic/tox.ini (pid 6031)
E       using tox-3.13.3.dev6+g1e6430c from /tmp/tox/.tox/pypy/site-packages/tox/__init__.pyc (pid 6031)
E       .package start: getenv /tmp/pytest-of-asottile/pytest-56/popen-gw6/test_package_setuptools0/magic/.tox/.package
E       .package cannot reuse: no previous config /tmp/pytest-of-asottile/pytest-56/popen-gw6/test_package_setuptools0/magic/.tox/.package/.tox-config1
E       .package create: /tmp/pytest-of-asottile/pytest-56/popen-gw6/test_package_setuptools0/magic/.tox/.package
E       .package uses None
E       ERROR: InterpreterNotFound: python
E       .package finish: getenv /tmp/pytest-of-asottile/pytest-56/popen-gw6/test_package_setuptools0/magic/.tox/.package after 0.00 seconds
```

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -35,6 +35,10 @@ jobs:
         image: [linux, windows, macOs]
       py27:
         image: [linux, windows, macOs]
+      pypy:
+        image: [linux]
+      pypy3:
+        image: [linux]
       py36:
         image: [linux, windows, macOs]
       py35:

--- a/docs/changelog/1378.bugfix.rst
+++ b/docs/changelog/1378.bugfix.rst
@@ -1,0 +1,1 @@
+Fix ``current_tox_py`` for ``pypy`` / ``pypy3`` - by :user:`asottile`

--- a/src/tox/_pytestplugin.py
+++ b/src/tox/_pytestplugin.py
@@ -577,7 +577,7 @@ def mock_venv(monkeypatch):
 def current_tox_py():
     """generate the current (test runners) python versions key
     e.g. py37 when running under Python 3.7"""
-    return "py{}".format("".join(str(i) for i in sys.version_info[0:2]))
+    return "{}{}{}".format("pypy" if tox.INFO.IS_PYPY else "py", *sys.version_info)
 
 
 def pytest_runtest_setup(item):

--- a/tests/integration/test_package_int.py
+++ b/tests/integration/test_package_int.py
@@ -14,17 +14,19 @@ def test_package_setuptools(initproj, cmd):
     initproj(
         "magic-0.1",
         filedefs={
-            "tox.ini": """
-                    [tox]
-                    isolated_build = true
-                    [testenv:.package]
-                    basepython = python
-                """,
-            "pyproject.toml": """
-                    [build-system]
-                    requires = ["setuptools >= 35.0.2", "setuptools_scm >= 2.0.0, <3"]
-                    build-backend = "setuptools.build_meta"
-                    """,
+            "tox.ini": """\
+                [tox]
+                isolated_build = true
+                [testenv:.package]
+                basepython = {}
+            """.format(
+                sys.executable
+            ),
+            "pyproject.toml": """\
+                [build-system]
+                requires = ["setuptools >= 35.0.2", "setuptools_scm >= 2.0.0, <3"]
+                build-backend = "setuptools.build_meta"
+            """,
         },
     )
     run(cmd, "magic-0.1.tar.gz")
@@ -37,26 +39,28 @@ def test_package_flit(initproj, cmd):
     initproj(
         "magic-0.1",
         filedefs={
-            "tox.ini": """
-                    [tox]
-                    isolated_build = true
-                    [testenv:.package]
-                    basepython = python
-                """,
-            "pyproject.toml": """
-                    [build-system]
-                    requires = ["flit"]
-                    build-backend = "flit.buildapi"
+            "tox.ini": """\
+                [tox]
+                isolated_build = true
+                [testenv:.package]
+                basepython = {}
+            """.format(
+                sys.executable
+            ),
+            "pyproject.toml": """\
+                [build-system]
+                requires = ["flit"]
+                build-backend = "flit.buildapi"
 
-                    [tool.flit.metadata]
-                    module = "magic"
-                    author = "Happy Harry"
-                    author-email = "happy@harry.com"
-                    home-page = "https://github.com/happy-harry/is"
-                    requires = [
-                        "tox",
-                    ]
-                    """,
+                [tool.flit.metadata]
+                module = "magic"
+                author = "Happy Harry"
+                author-email = "happy@harry.com"
+                home-page = "https://github.com/happy-harry/is"
+                requires = [
+                    "tox",
+                ]
+            """,
             ".gitignore": ".tox",
         },
         add_missing_setup_py=False,
@@ -78,24 +82,25 @@ def test_package_poetry(initproj, cmd):
     initproj(
         "magic-0.1",
         filedefs={
-            "tox.ini": """
-                    [tox]
-                    isolated_build = true
-                    [testenv:.package]
-                    basepython = python
-                """,
-            "pyproject.toml": """
-                    [build-system]
-                    requires = ["poetry>=0.12"]
-                    build-backend = "poetry.masonry.api"
+            "tox.ini": """\
+                [tox]
+                isolated_build = true
+                [testenv:.package]
+                basepython = {}
+            """.format(
+                sys.executable
+            ),
+            "pyproject.toml": """\
+                [build-system]
+                requires = ["poetry>=0.12"]
+                build-backend = "poetry.masonry.api"
 
-                    [tool.poetry]
-                    name = "magic"
-                    version = "0.1.0"
-                    description = ""
-                    authors = ["Name <email@email.com>"]
-
-                    """,
+                [tool.poetry]
+                name = "magic"
+                version = "0.1.0"
+                description = ""
+                authors = ["Name <email@email.com>"]
+            """,
             ".gitignore": ".tox",
         },
         add_missing_setup_py=False,

--- a/tests/unit/config/test_config.py
+++ b/tests/unit/config/test_config.py
@@ -1909,17 +1909,21 @@ class TestConfigTestEnv:
 
     def test_default_factors_conflict(self, newconfig, capsys):
         with pytest.warns(UserWarning, match=r"conflicting basepython .*"):
+            exe = "pypy3" if tox.INFO.IS_PYPY else "python3"
+            env = "pypy27" if tox.INFO.IS_PYPY else "py27"
             config = newconfig(
-                """
+                """\
                 [testenv]
-                basepython=python3
-                [testenv:py27]
+                basepython={}
+                [testenv:{}]
                 commands = python --version
-            """
+                """.format(
+                    exe, env
+                )
             )
         assert len(config.envconfigs) == 1
-        envconfig = config.envconfigs["py27"]
-        assert envconfig.basepython == "python3"
+        envconfig = config.envconfigs[env]
+        assert envconfig.basepython == exe
 
     def test_default_factors_conflict_lying_name(
         self, newconfig, capsys, tmpdir, recwarn, monkeypatch

--- a/tests/unit/interpreters/test_interpreters.py
+++ b/tests/unit/interpreters/test_interpreters.py
@@ -28,6 +28,7 @@ def create_interpreters_instance():
     return Interpreters(hook=pm.hook)
 
 
+@pytest.mark.skipif(tox.INFO.IS_PYPY, reason="testing cpython interpreter discovery")
 def test_tox_get_python_executable():
     class envconfig:
         basepython = sys.executable

--- a/tests/unit/test_venv.py
+++ b/tests/unit/test_venv.py
@@ -446,8 +446,8 @@ def test_install_command_not_installed_bash(newmocksession):
 
 
 def test_install_python3(newmocksession):
-    if not py.path.local.sysfind("python3"):
-        pytest.skip("needs python3")
+    if not py.path.local.sysfind("python3") or tox.INFO.IS_PYPY:
+        pytest.skip("needs cpython3")
     mocksession = newmocksession(
         [],
         """\

--- a/tests/unit/test_z_cmdline.py
+++ b/tests/unit/test_z_cmdline.py
@@ -784,11 +784,13 @@ def test_notest(initproj, cmd):
     initproj(
         "example123",
         filedefs={
-            "tox.ini": """
-        # content of: tox.ini
-        [testenv:py26]
-        basepython=python
-    """
+            "tox.ini": """\
+            # content of: tox.ini
+            [testenv:py26]
+            basepython={}
+            """.format(
+                sys.executable
+            )
         },
     )
     result = cmd("-v", "--notest")

